### PR TITLE
Hotfix/2657 prevent share overwrite

### PR DESCRIPTION
**Pull request by ckieschnick**
Added warnings and prevented export when exports are conflicting, added warning when imports are using the same source.

Fixed an additional bug which prevented the usage of exports when a database was newly created. 

Addresses an issue discussed in #2657.

## Type of change
- ✅ Bug fix (non-breaking change which fixes an issue)

## Description and Context
Creation of a new database did not trigger the export even when needed references were defined. 
In case of multiple usage of the same export/import container, there was no indication about that. Now the import/export and configuration gives warnings/errors about those cases.

## Testing strategy
Manually tested.

## Points to discuss
Currently the checking uses the entered paths. We may extend the checking to use canonical paths which may reduce (but not eliminate) the error rate. 
An alternative (but does not prevent export!) would be to open and check the container for an export flag in the public data which indicates if the container was already written within the "transaction" and to abort the following exports for this container. I'm not sure if this would be a better solution to the problem since we do export data partially and (for now) there is no guarantee that the order of exports is deterministic.

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

### CODE REVIEW

File: src/gui/DatabaseWidget.cpp [hunk 1]
```diff
@@ -370,7 +370,9 @@ void DatabaseWidget::replaceDatabase(QSharedPointer<Database> db)
     // TODO: instead of increasing the ref count temporarily, there should be a clean
     // break from the old database. Without this crashes occur due to the change
     // signals triggering dangling pointers.
+#if defined(WITH_XC_KEESHARE)
     auto oldDb = m_db;
+#endif
```
Comment (droidmonkey) [inline src/gui/DatabaseWidget.cpp]: Don't guard this, it is required to maintain the database pointer during the life of this function. Using it with KeeShare is an added benefit (see the TODO above it).

File: src/gui/DatabaseWidget.cpp [hunk 2]
```diff
@@ -1548,7 +1550,11 @@ bool DatabaseWidget::saveAs()
             // Ensure we don't recurse back into this function
             m_db->setReadOnly(false);
             m_db->setFilePath(newFilePath);
-
+#ifdef WITH_XC_KEESHARE
+            // We need to reregister the database to allow exports
+            // from a newly created database
+            KeeShare::instance()->connectDatabase(m_db, m_db);
```
Comment (droidmonkey) [inline src/gui/DatabaseWidget.cpp]: I don't necessarily like passing the "old" database pointer into connectDatabase. The keeshare observer should listen for the databases' ```QObject::destroyed``` signal and handle any necessary cleanup. It is unnecessary to disconnect signal/slots from an object being deleted.

File: src/gui/group/EditGroupWidget.h [hunk 3]
```diff
@@ -43,7 +43,7 @@ class IEditGroupPage
     virtual QString name() = 0;
     virtual QIcon icon() = 0;
     virtual QWidget* createWidget() = 0;
-    virtual void set(QWidget* widget, Group* tempoaryGroup) = 0;
+    virtual void set(QWidget* widget, Group* tempoaryGroup, const Database* database) = 0;
```
Comment (droidmonkey) [inline src/gui/group/EditGroupWidget.h]: This should be a ```QSharedPointer<Database>```, we do not want to pass around bare database pointers even if they are const.

File: src/keeshare/group/EditGroupPageKeeShare.h [hunk 4]
```diff
@@ -30,7 +30,7 @@ class EditGroupPageKeeShare : public IEditGroupPage
     QString name() override;
     QIcon icon() override;
     QWidget* createWidget() override;
-    void set(QWidget* widget, Group* temporaryGroup) override;
+    void set(QWidget* widget, Group* temporaryGroup, const Database* database) override;
```
Comment (droidmonkey) [inline src/keeshare/group/EditGroupPageKeeShare.h]: Same as above, ```QSharedPointer<Database>```

File: src/keeshare/group/EditGroupWidgetKeeShare.h [hunk 5]
```diff
@@ -37,7 +37,7 @@ class EditGroupWidgetKeeShare : public QWidget
     explicit EditGroupWidgetKeeShare(QWidget* parent = nullptr);
     ~EditGroupWidgetKeeShare();
 
-    void setGroup(Group* temporaryGroup);
+    void setGroup(Group* temporaryGroup, const Database* database);
```
Comment (droidmonkey) [inline src/keeshare/group/EditGroupWidgetKeeShare.h]: Same as above, ```QSharedPointer<Database>```

File: src/gui/group/EditGroupWidget.h (hunk 3 above)
Comment (droidmonkey) [reply src/gui/group/EditGroupWidget.h]: When you are passing the database you can use: ```std::move(m_db)``` to prevent a copy

**Comment by ckieschnick** [comment]
Concerning the shared pointer used by the `EditGroupWidget` and `EditEntryWidget` - is it possible that this pointer does extend the lifetime of the database pointer unnecessarily?

**Comment by droidmonkey** [comment]
Yes that is possible. You could use a QWeakPointer in this case.

**Comment by ckieschnick** [comment]
I'd had a closer look at the source code. It seems, that there are a number of `QSharedPointer` (introduced before this pull request) which may also influence the lifetime of the object. `EditWidgetIcons` uses currently the same pattern. 

Since it is already used without negative effects, I would suggest to keep the shared pointer for now and check if the change to `QWeakPointer` is necessary in a separate ticket. Maybe it can be combined with a general refactoring task to populate the settings widgets like `EditEntryWidget` with the specialized widgets fulfilling a simple interface (like already introduced for `EditGroupWidget` with *KeeShare* and `DatabaseSettingsDialog`).

**Pull request merged**